File: mobile/examples/java/hello_world/MainActivity.java
package io.envoyproxy.envoymobile.helloenvoy;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.Log;
import io.envoyproxy.envoymobile.AndroidEngineBuilder;
import io.envoyproxy.envoymobile.Counter;
import io.envoyproxy.envoymobile.Engine;
import io.envoyproxy.envoymobile.Element;
import io.envoyproxy.envoymobile.LogLevel;
import io.envoyproxy.envoymobile.RequestHeaders;
import io.envoyproxy.envoymobile.RequestHeadersBuilder;
import io.envoyproxy.envoymobile.RequestMethod;
import io.envoyproxy.envoymobile.ResponseHeaders;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import io.envoyproxy.envoymobile.shared.Failure;
import io.envoyproxy.envoymobile.shared.ResponseRecyclerViewAdapter;
import io.envoyproxy.envoymobile.shared.Success;
import kotlin.Unit;

import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class MainActivity extends Activity {
  private static final String TAG = "MainActivity";
  private static final String REQUEST_HANDLER_THREAD_NAME = "hello_envoy_java";
  private static final String ENVOY_SERVER_HEADER = "server";
  private static final String REQUEST_AUTHORITY = "api.lyft.com";
  private static final String REQUEST_PATH = "/ping";
  private static final String REQUEST_SCHEME_HTTP = "http";
  private static final String REQUEST_SCHEME_HTTPS = "https";
  private static final Set<String> FILTERED_HEADERS =
      Set.of("server", "filter-demo", "x-envoy-upstream-service-time");

  private Engine engine;
  private RecyclerView recyclerView;

  private HandlerThread thread = new HandlerThread(REQUEST_HANDLER_THREAD_NAME);
  private ResponseRecyclerViewAdapter viewAdapter;
  private Boolean clear_text = true;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    engine = new AndroidEngineBuilder(getApplication())
                 .setLogLevel(LogLevel.DEBUG)
                 .setLogger((level, message) -> {
                   Log.d(TAG, message);
                   return null;
                 })
                 .setOnEngineRunning(() -> {
                   Log.d(TAG, "Envoy async internal setup completed");
                   return null;
                 })
                 .enablePlatformCertificatesValidation(true)
                 .build();

    recyclerView = (RecyclerView)findViewById(R.id.recycler_view);
    recyclerView.setLayoutManager(new LinearLayoutManager(this));

    viewAdapter = new ResponseRecyclerViewAdapter();
    recyclerView.setAdapter(viewAdapter);
    DividerItemDecoration dividerItemDecoration =
        new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL);
    recyclerView.addItemDecoration(dividerItemDecoration);
    thread.start();

    // Run a request loop and record stats until the application exits.
    final Handler handler = new Handler(thread.getLooper());
    handler.postDelayed(new Runnable() {
      @Override
      public void run() {
        makeRequest();
        recordStats();
        // Make a call and record stats again
        handler.postDelayed(this, TimeUnit.SECONDS.toMillis(1));
      }
    }, TimeUnit.SECONDS.toMillis(1));
  }

  protected void onDestroy() {
    super.onDestroy();
    thread.quit();
  }

  private void makeRequest() {
    // Note: this request will use an http/1.1 stream for the upstream request.
    // The Kotlin example uses h2. This is done on purpose to test both paths in
    // end-to-end tests in CI.
    String scheme = clear_text ? REQUEST_SCHEME_HTTP : REQUEST_SCHEME_HTTPS;
    RequestHeaders requestHeaders =
        new RequestHeadersBuilder(RequestMethod.GET, scheme, REQUEST_AUTHORITY, REQUEST_PATH)
            .build();
    engine.streamClient()
        .newStreamPrototype()
        .setOnResponseHeaders((responseHeaders, endStream, ignored) -> {
          Integer status = responseHeaders.getHttpStatus();
          String message = "received headers with status " + status + " for " + scheme + " request";

          StringBuilder sb = new StringBuilder();
          for (Map.Entry<String, List<String>> entry :
               responseHeaders.caseSensitiveHeaders().entrySet()) {
            String name = entry.getKey();
            if (FILTERED_HEADERS.contains(name)) {
              sb.append(name).append(": ").append(String.join(", ", entry.getValue())).append("\n");
            }
          }
          String headerText = sb.toString();

          Log.d(TAG, message);
          if ((scheme == REQUEST_SCHEME_HTTP && status == 301) ||
              (scheme == REQUEST_SCHEME_HTTPS && status == 200)) {
            // The server returns 301 to http request and 200 to https request.
            String serverHeaderField = responseHeaders.value(ENVOY_SERVER_HEADER).get(0);
            recyclerView.post(() -> viewAdapter.add(new Success(message, headerText)));
          } else {
            recyclerView.post(() -> viewAdapter.add(new Failure(message)));
          }
          return Unit.INSTANCE;
        })
        .setOnError((error, ignored) -> {
          String message = "failed with error after " + error.getAttemptCount() +
                           " attempts: " + error.getMessage() + " for " + scheme + " request";
          Log.d(TAG, message);
          recyclerView.post(() -> viewAdapter.add(new Failure(message)));
          return Unit.INSTANCE;
        })
        .start(Executors.newSingleThreadExecutor())
        .sendHeaders(requestHeaders, /* endStream= */ true, /* idempotent= */ false);

    clear_text = !clear_text;
  }

  private void recordStats() {
    final Counter counter = engine.pulseClient().counter(new Element("foo"), new Element("bar"),
                                                         new Element("counter"));
    counter.increment(1);
    counter.increment(5);
  }
}

File: mobile/examples/kotlin/hello_world/MainActivity.kt
package io.envoyproxy.envoymobile.helloenvoykotlin

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.util.Log
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.envoyproxy.envoymobile.AndroidEngineBuilder
import io.envoyproxy.envoymobile.Element
import io.envoyproxy.envoymobile.Engine
import io.envoyproxy.envoymobile.LogLevel
import io.envoyproxy.envoymobile.RequestHeadersBuilder
import io.envoyproxy.envoymobile.RequestMethod
import io.envoyproxy.envoymobile.shared.Failure
import io.envoyproxy.envoymobile.shared.ResponseRecyclerViewAdapter
import io.envoyproxy.envoymobile.shared.Success
import java.io.IOException
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

private const val REQUEST_HANDLER_THREAD_NAME = "hello_envoy_kt"
private const val REQUEST_AUTHORITY = "api.lyft.com"
private const val REQUEST_PATH = "/ping"
private const val REQUEST_SCHEME = "https"
private val FILTERED_HEADERS =
  setOf(
    "server",
    "filter-demo",
    "buffer-filter-demo",
    "async-filter-demo",
    "x-envoy-upstream-service-time"
  )

/** The main activity of the app. */
class MainActivity : Activity() {
  private val thread = HandlerThread(REQUEST_HANDLER_THREAD_NAME)
  private lateinit var recyclerView: RecyclerView
  private lateinit var viewAdapter: ResponseRecyclerViewAdapter
  private lateinit var engine: Engine

  @Suppress("MaxLineLength")
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    engine =
      AndroidEngineBuilder(application)
        .addLogLevel(LogLevel.DEBUG)
        .enableProxying(true)
        .addPlatformFilter(::DemoFilter)
        .addPlatformFilter(::BufferDemoFilter)
        .addPlatformFilter(::AsyncDemoFilter)
        .addNativeFilter(
          "envoy.filters.http.buffer",
          "{\"@type\":\"type.googleapis.com/envoy.extensions.filters.http.buffer.v3.Buffer\",\"max_request_bytes\":5242880}"
        )
        .addStringAccessor("demo-accessor", { "PlatformString" })
        .setOnEngineRunning { Log.d("MainActivity", "Envoy async internal setup completed") }
        .setEventTracker({
          for (entry in it.entries) {
            Log.d("MainActivity", "Event emitted: ${entry.key}, ${entry.value}")
          }
        })
        .setLogger { Log.d("MainActivity", it) }
        .build()

    recyclerView = findViewById(R.id.recycler_view) as RecyclerView
    recyclerView.layoutManager = LinearLayoutManager(this)

    viewAdapter = ResponseRecyclerViewAdapter()
    recyclerView.adapter = viewAdapter
    val dividerItemDecoration =
      DividerItemDecoration(recyclerView.context, DividerItemDecoration.VERTICAL)
    recyclerView.addItemDecoration(dividerItemDecoration)
    thread.start()
    val handler = Handler(thread.looper)

    // Run a request loop and record stats until the application exits.
    handler.postDelayed(
      object : Runnable {
        override fun run() {
          try {
            makeRequest()
            recordStats()
          } catch (e: IOException) {
            Log.d("MainActivity", "exception making request or recording stats", e)
          }

          // Make a call and report stats again
          handler.postDelayed(this, TimeUnit.SECONDS.toMillis(1))
        }
      },
      TimeUnit.SECONDS.toMillis(1)
    )
  }

  override fun onDestroy() {
    super.onDestroy()
    thread.quit()
  }

  private fun makeRequest() {
    // Note: this request will use an h2 stream for the upstream request due to https scheme.
    // The Java example uses http so http/1.1. This is done on purpose to test both paths in
    // end-to-end tests in CI.
    val requestHeaders =
      RequestHeadersBuilder(RequestMethod.GET, REQUEST_SCHEME, REQUEST_AUTHORITY, REQUEST_PATH)
        .build()
    engine
      .streamClient()
      .newStreamPrototype()
      .setOnResponseHeaders { responseHeaders, _, _ ->
        val status = responseHeaders.httpStatus ?: 0L
        val message = "received headers with status $status"

        val sb = StringBuilder()
        for ((name, value) in responseHeaders.caseSensitiveHeaders()) {
          if (name in FILTERED_HEADERS) {
            sb.append(name).append(": ").append(value.joinToString()).append("\n")
          }
        }
        val headerText = sb.toString()

        Log.d("MainActivity", message)
        responseHeaders.value("filter-demo")?.first()?.let { filterDemoValue ->
          Log.d("MainActivity", "filter-demo: $filterDemoValue")
        }

        if (status == 200) {
          recyclerView.post { viewAdapter.add(Success(message, headerText)) }
        } else {
          recyclerView.post { viewAdapter.add(Failure(message)) }
        }
      }
      .setOnError { error, _ ->
        val attemptCount = error.attemptCount ?: -1
        val message = "failed with error after $attemptCount attempts: ${error.message}"
        Log.d("MainActivity", message)
        recyclerView.post { viewAdapter.add(Failure(message)) }
      }
      .start(Executors.newSingleThreadExecutor())
      .sendHeaders(requestHeaders, true)
  }

  private fun recordStats() {
    val counter = engine.pulseClient().counter(Element("foo"), Element("bar"), Element("counter"))
    counter.increment()
    counter.increment(5)
  }
}
